src/lib.rs

```diff
@@ -89,27 +93,55 @@ impl<T> ReusableBoxFuture<T> {
     where
         F: Future<Output = T> + Send + 'static,
     {
-        // Drop the existing future, catching any panics.
-        let result = panic::catch_unwind(AssertUnwindSafe(|| {
-            ptr::drop_in_place(self.boxed.as_ptr());
-        }));
-
-        // Overwrite the future behind the pointer. This is safe because the
-        // allocation was allocated with the same size and alignment as the type F.
-        let self_ptr: *mut F = self.boxed.as_ptr() as *mut F;
-        ptr::write(self_ptr, future);
-
-        // Update the vtable of self.boxed. The pointer is not null because we
-        // just got it from self.boxed, which is not null.
-        self.boxed = NonNull::new_unchecked(self_ptr);
-
-        // If the old future's destructor panicked, resume unwinding.
-        match result {
-            Ok(()) => {}
-            Err(payload) => {
-                panic::resume_unwind(payload);
+        struct SetLayout<'a, F, T>
+        where
+            F: Future<Output = T> + Send + 'static,
+        {
+            rbf: &'a mut ReusableBoxFuture<T>,
+            new_future: ManuallyDrop<F>,
+        }
+
+        impl<'a, F, T> SetLayout<'a, F, T>
+        where
+            F: Future<Output = T> + Send + 'static,
+        {
+            fn apply(self) {
+                unsafe {
+                    // Drop the existing future.
+                    ptr::drop_in_place(self.rbf.boxed.as_ptr());
```

Comment:
I just realized we don't even need the method. We can just do `ptr::drop_in_place(set_layout.rbf.boxed.as_ptr())` instead of calling a method.